### PR TITLE
Add routes command

### PR DESCRIPTION
**Pull request by kiwiroy**
Contributing to the routes command.

### CODE REVIEW

File: lib/cli/routes.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  // Methods
+  row.push(route.methods.length === 0 ? '*' : route.methods.map(method => method.toUpperCase()).join(','));
+
+  // Name
+  const name = route.defaultName;
+  row.push(route.hasCustomName() ? `"${name}"` : name);
+
+  // Regex (verbose)
+  const pattern = route.pattern;
+  pattern.match('/', {isEndpoint: route.is_endpoint, partial: route.partial});
```
Comment (kraih) [inline lib/cli/routes.js:48]: Doesn't exist!

File: lib/router/route.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ export default class Route {
         child.defaultName = arg.replace(/[^0-9a-z]+/gi, '_').replace(/^_|_$/g, '');
         child.pattern.parse(arg);
       } else if (arg instanceof Array) {
-        child.methods = this.methods.concat(arg);
+        child.methods = child.methods.concat(arg);
```
Comment (kraih) [inline lib/router/route.js:29]: Looks like an unrelated bug fix.

File: lib/router/route.js [hunk 3]
```diff
@@ -46,6 +46,10 @@ export default class Route {
     return this.any(['GET'], ...args);
   }
 
+  hasCustomName () {
+    return this.customName !== undefined;
+  }
```
Comment (kraih) [inline lib/router/route.js:51]: Not needed at all.

File: test/support/command-app/index.js [hunk 4]
```diff
@@ -4,4 +4,7 @@ export const app = mojo();
 
 app.any('/', ctx => ctx.render({text: 'Hello Mojo!'}));
 
+const foo = app.get('/foo', ctx => ctx.render({text: 'foo'}));
+foo.post('/bar', ctx => ctx.render({text: 'bar'}));
```
Comment (kraih) [inline test/support/command-app/index.js:8]: The closures are not needed just to test the route.

**Review by kraih — CHANGES_REQUESTED** [review]

File: lib/cli/routes.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+routesCommand.description = 'Show available routes';
+routesCommand.usage = `Usage: APPLICATION routes [OPTIONS]
+
+  ./myapp.js routes
+  ./myapp.js routes -v
+
+Options:
+  -h, --help            Show this summary of available options
+  -v, --verbose         Print additional details about routes, flags indicate
+                        C=Conditions, P=Partial, U=Under and W=WebSocket
```
[the remark this answers is not in the document]
Comment (kraih) [reply lib/cli/routes.js:23]: Also includes features that do not exist in mojo.js.

File: lib/router/route.js (hunk 2 above)
Comment (kraih) [reply lib/router/route.js:29]: Also doesn't look quite right to me. 🤔

File: lib/cli/routes.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+`;
+
+function _walk (route, depth, rows, verbose) {
+  const prefix = ' '.repeat(depth * 2);
+  const row = [prefix + (route.pattern.unparsed || '/') + (depth ? '+' : '')];
```
Comment (kraih) [inline lib/cli/routes.js:28]: Use explicit type checks whenever possible.

Comment (kraih) [reply lib/cli/routes.js:28]: This looks like a case for `??`.

File: lib/router/route.js (hunk 2 above)
Comment (kraih) [reply lib/router/route.js:29]: Since it's not related to the rest of the PR, i've fixed this already.

**Comment by kraih** [comment]
Just leave out the flags from the verbose output.

File: lib/cli/routes.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import nopt from 'nopt';
+import {tablify} from '../util.js';
+
+export default async function routesCommand (app, args) {
+  const parsed = nopt({verbose: Boolean},
+    {v: '--verbose'}, args, 1);
```
[the remark this answers is not in the document]
Comment (kiwiroy) [reply lib/cli/routes.js]: fixed

File: lib/cli/routes.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+function _walk (route, depth, rows, verbose) {
+  const prefix = ' '.repeat(depth * 2);
+  const row = [prefix + (route.pattern.unparsed || '/') + (depth ? '+' : '')];
+  rows.push(row);
+
+  // Methods
+  row.push(route.methods.length === 0 ? '*' : route.methods.map(method => method.toUpperCase()).join(','));
+
+  // Name
+  row.push(typeof route.customName !== 'undefined' ? `"${route.customName}"` : route.defaultName);
```
Comment (kraih) [inline lib/cli/routes.js:31]: Why is that better than `route.customName !== undefined`?

File: lib/cli/routes.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  // Regex (verbose)
+  const pattern = route.pattern;
+  pattern.match('/', {});
```
Comment (kraih) [inline lib/cli/routes.js:35]: Why the empty object?

File: lib/cli/routes.js (hunk 8 above)
Comment (oetiker) [reply lib/cli/routes.js:31]: in some js universes `undefined` is a variable which can have a value

Comment (kraih) [reply lib/cli/routes.js:31]: 🙄

File: lib/cli/routes.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  ./myapp.js routes -v
+
+Options:
+  -h, --help      Show this summary of available options
+  -v, --verbose   Print additional details about routes
+`;
+
+function _walk (route, depth, rows, verbose) {
+  const prefix = ' '.repeat(depth * 2);
+  const row = [prefix + (route.pattern.unparsed || '/') + (depth ? '+' : '')];
```
Comment (kraih) [inline lib/cli/routes.js:24]: That `+` handling looks broken.

**Review by kraih — CHANGES_REQUESTED** [review]

File: lib/cli/routes.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+  // Methods
+  row.push(route.methods.length === 0 ? '*' : route.methods.map(method => method.toUpperCase()).join(','));
+
+  // Name
+  row.push(route.customName !== undefined ? `"${route.customName}"` : route.defaultName);
+
+  // Regex (verbose)
+  const pattern = route.pattern;
+  pattern.match('/');
```
[the remark this answers is not in the document]
Comment (kraih) [reply lib/cli/routes.js:35]: Noticeable in the output if you add a file extension constraint.

**Comment by kraih** [comment]
Thanks!
